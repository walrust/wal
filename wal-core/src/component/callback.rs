use std::{hash::Hash, rc::Rc};

/// Callback is a wrapper around a function that is used to send messages to the [Component](../trait.Component.html).
///
/// `IN` is the type of the input of the wrapped function.
/// Meaning if we would like to send a message to the [component](../trait.Component.html) we need to provide input of type `IN`.
/// Callback should be used for defining child to parent [component](../trait.Component.html) communication and subscribing
/// to HTML [events](../../events/index.html).
pub struct Callback<IN> {
    wrapper: Rc<dyn Fn(IN)>,
}

impl<IN> Callback<IN> {
    /// Function that creates a new instance of the Callback.
    /// Using a Callback defined by this function will not send [messages](../trait.Component.html#associatedtype.Message) to the [component](../trait.Component.html).
    /// It should be used only if the Callback is to send [messages](../trait.Component.html#associatedtype.Message)
    /// to the father [component](../trait.Component.html) and this logic should be contained in `wrapper` argument function.
    /// If sending the [message](../trait.Component.html#associatedtype.Message) to the current [component](../trait.Component.html)
    /// is the goal [create_callback](../behavior/trait.Behavior.html#tymethod.create_callback) function should be used.
    pub fn new<F>(wrapper: F) -> Self
    where
        F: Fn(IN) + 'static,
    {
        Callback {
            wrapper: Rc::new(wrapper),
        }
    }

    /// Function that calls the [Callback] using the provided `input`.
    pub fn emit(&self, input: IN) {
        (self.wrapper)(input);
    }
}

impl<IN> Hash for Callback<IN> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        let ptr = self.wrapper.as_ref() as *const dyn Fn(IN);
        ptr.hash(state);
    }
}

impl<IN> Clone for Callback<IN> {
    fn clone(&self) -> Self {
        Self {
            wrapper: self.wrapper.clone(),
        }
    }
}
