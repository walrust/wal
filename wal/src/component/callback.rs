use std::hash::{Hash, Hasher};

/// Once created callback for child shouldnt change, the change isnt registered in runtime
pub struct Callback<IN> {
    wrapper: Box<dyn Fn(IN)>,
}

impl<IN> Hash for Callback<IN> {
    fn hash<H: Hasher>(&self, _state: &mut H) {
    }
}

impl<IN> Callback<IN> {
    pub fn new<F>(wrapper: F) -> Self
    where
        F: Fn(IN) + 'static,
    {
        Callback {
            wrapper: Box::new(wrapper),
        }
    }

    pub fn emit(&self, input: IN) {
        (self.wrapper)(input);
    }
}
