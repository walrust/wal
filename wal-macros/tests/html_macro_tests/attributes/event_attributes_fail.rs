use wal::component::callback::Callback;
use wal_macros::html;

fn main() {
    html! { <div onclick></div> };
    html! { <div onclick=></div> };
    html! { <div onclick {Callback::new(|_event: web_sys::MouseEvent)| {}}></div> };
    html! {
        <div
            onclick={Callback::new(|_event: web_sys::MouseEvent| {})}
            onclick={Callback::new(|_event: web_sys::MouseEvent| {})}>
        </div>
    };
    html! { <div onclick= ondblclick={Callback::new(|_event: web_sys::MouseEvent| {})}></div> };
    html! { <div onclick={Callback::new(|_event: web_sys::Event| {})}></div> };
    html! { <div onclick={Callback::new(|_event: web_sys::DragEvent| {})}></div> };
    html! { <div onclick="value"></div> };
    html! { <div onclick={}></div> };
    html! { <div onclick={"value"}></div> };
}
