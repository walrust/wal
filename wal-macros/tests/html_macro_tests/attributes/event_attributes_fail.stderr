error: expected `=`
 --> tests/html_macro_tests/attributes/event_attributes_fail.rs:5:25
  |
5 |     html! { <div onclick></div> };
  |                         ^

error: Expected a literal or an expression block
 --> tests/html_macro_tests/attributes/event_attributes_fail.rs:6:26
  |
6 |     html! { <div onclick=></div> };
  |                          ^

error: expected `=`
 --> tests/html_macro_tests/attributes/event_attributes_fail.rs:7:26
  |
7 |     html! { <div onclick {Callback::new(|_event: web_sys::MouseEvent)| {}}></div> };
  |                          ^

error: Duplicate event attribute `onclick`
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:11:13
   |
11 |             onclick={Callback::new(|_event: web_sys::MouseEvent| {})}>
   |             ^^^^^^^

error: Expected a literal or an expression block
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:14:27
   |
14 |     html! { <div onclick= ondblclick={Callback::new(|_event: web_sys::MouseEvent| {})}></div> };
   |                           ^^^^^^^^^^

error: Expected an expression block for event attribute `onclick`
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:17:18
   |
17 |     html! { <div onclick="value"></div> };
   |                  ^^^^^^^

error: Expected a non-empty expression block
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:18:26
   |
18 |     html! { <div onclick={}></div> };
   |                          ^^

error[E0308]: mismatched types
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:15:27
   |
15 |     html! { <div onclick={Callback::new(|_event: web_sys::Event| {})}></div> };
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Callback<MouseEvent>`, found `Callback<Event>`
   |
   = note: expected struct `Callback<MouseEvent>`
              found struct `Callback<Event>`

error[E0308]: mismatched types
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:16:27
   |
16 |     html! { <div onclick={Callback::new(|_event: web_sys::DragEvent| {})}></div> };
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Callback<MouseEvent>`, found `Callback<DragEvent>`
   |
   = note: expected struct `Callback<MouseEvent>`
              found struct `Callback<DragEvent>`

error[E0308]: mismatched types
  --> tests/html_macro_tests/attributes/event_attributes_fail.rs:19:27
   |
19 |     html! { <div onclick={"value"}></div> };
   |                           ^^^^^^^ expected `Callback<MouseEvent>`, found `&str`
   |
   = note: expected struct `Callback<MouseEvent>`
           found reference `&'static str`
